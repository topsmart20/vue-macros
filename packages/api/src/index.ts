export * from '@vue-macros/common'

export * from './vue'
export * from './resolve'
export * from './ts'
export * from './utils'
