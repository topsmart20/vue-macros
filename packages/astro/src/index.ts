import { type AstroIntegration, type ViteUserConfig } from 'astro'
import { type Options, resolveOptions } from 'unplugin-vue-macros'
import { type Plugin } from 'vite'
import VueMacros from 'unplugin-vue-macros/vite'

export type VueMacrosOptions = Options

function findPluginAndRemove(
  name: string,
  plugins: ViteUserConfig['plugins']
): Plugin | undefined {
  const idx = plugins!.findIndex(
    (plugin) => plugin && 'name' in plugin && plugin.name === name
  )
  if (idx === -1) return
  const plugin = plugins![idx]
  plugins!.splice(idx, 1)
  return plugin as any
}

export default function (options?: VueMacrosOptions): AstroIntegration {
  return {
    name: '@vue-macros/astro',
    hooks: {
      'astro:config:setup': ({ config }) => {
        const resolvedOptions = resolveOptions(options ?? {})
        const vue = findPluginAndRemove('vite:vue', config.vite.plugins)
        const vueJsx = findPluginAndRemove('vite:vue-jsx', config.vite.plugins)

        config.vite.plugins?.push(
          VueMacros({
            ...resolvedOptions,
            plugins: {
              vue,
              vueJsx,
            },
          })
        )
      },
    },
  }
}
