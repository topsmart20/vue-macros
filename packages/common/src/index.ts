export * from './vue'
export * from './ast'
export * from './constants'
export * from './unplugin'
export { default as MagicString } from 'magic-string'
