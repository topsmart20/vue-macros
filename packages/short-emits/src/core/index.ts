import {
  MagicString,
  getTransformResult,
  isTs,
  parseSFC,
  resolveObjectKey,
  walkAST,
} from '@vue-macros/common'
import type {
  Identifier,
  Node,
  RestElement,
  TSMethodSignature,
  TSPropertySignature,
  TSType,
  TSTypeReference,
} from '@babel/types'

export function transformShortEmits(code: string, id: string) {
  if (!code.includes('SE') && !code.includes('ShortEmits')) return

  const sfc = parseSFC(code, id)
  const { scriptSetup, lang, getSetupAst } = sfc
  if (!scriptSetup || !isTs(lang)) return

  const offset = scriptSetup.loc.start.offset
  const ast = getSetupAst()!

  const nodes: {
    def: TSType
    type: TSTypeReference
  }[] = []

  walkAST<Node>(ast, {
    enter(node) {
      if (
        node.type === 'TSTypeReference' &&
        node.typeName.type === 'Identifier' &&
        ['SE', 'ShortEmits'].includes(node.typeName.name) &&
        node.typeParameters?.params[0].type
      ) {
        nodes.push({
          def: node.typeParameters.params[0],
          type: node,
        })
      }
    },
  })

  const s = new MagicString(code)

  function stringifyParams(params: Array<Identifier | RestElement>) {
    return params.length > 0 ? s.sliceNodes(params, { offset }) : ''
  }

  for (const { def, type } of nodes) {
    if (def.type !== 'TSTypeLiteral')
      throw new SyntaxError(
        `accepts object literal only: ${s.sliceNode(def, { offset })}`
      )

    // remove SE<...>
    s.remove(offset + type.start!, offset + def.start!)
    s.remove(offset + def.end!, offset + type.end!)

    for (const _member of def.members) {
      if (!['TSPropertySignature', 'TSMethodSignature'].includes(_member.type))
        throw new SyntaxError(
          `accepts method and property only: ${s.sliceNode(_member, {
            offset,
          })}`
        )

      const member = _member as TSPropertySignature | TSMethodSignature

      const key = resolveObjectKey(member.key, member.computed)
      let params = ''

      switch (member.type) {
        case 'TSPropertySignature': {
          if (
            !member.typeAnnotation ||
            !['TSTupleType', 'TSFunctionType'].includes(
              member.typeAnnotation.typeAnnotation.type
            )
          )
            throw new SyntaxError(
              `not supported: ${s.sliceNode(member, { offset })}`
            )

          switch (member.typeAnnotation.typeAnnotation.type) {
            case 'TSTupleType':
              params = `...args: ${s.sliceNode(
                member.typeAnnotation.typeAnnotation,
                { offset }
              )}`
              break
            case 'TSFunctionType':
              params = stringifyParams(
                member.typeAnnotation.typeAnnotation.parameters
              )
              break
          }
          break
        }

        case 'TSMethodSignature': {
          params = stringifyParams(member.parameters)
          break
        }
      }

      s.overwriteNode(
        member,
        `(evt: ${key}${params ? `, ${params}` : ''}): void`,
        { offset }
      )
    }
  }

  return getTransformResult(s, id)
}
