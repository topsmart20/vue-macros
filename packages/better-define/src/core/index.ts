import { MagicString, getTransformResult, parseSFC } from '@vue-macros/common'
import { analyzeSFC } from '@vue-macros/api'
import type { TSEmits, TSProps } from '@vue-macros/api'

export async function transformBetterDefine(
  code: string,
  id: string,
  isProduction?: boolean
) {
  const s = new MagicString(code)
  const sfc = parseSFC(code, id)
  if (!sfc.scriptSetup) return

  const offset = sfc.scriptSetup.loc.start.offset
  const result = await analyzeSFC(s, sfc)
  if (result.props) {
    await processProps(result.props)
  }
  if (result.emits) {
    processEmits(result.emits)
  }

  return getTransformResult(s, id)

  async function processProps(props: TSProps) {
    const runtimeDefs = await props.getRuntimeDefinitions()

    const runtimeDecls = `{\n  ${Object.entries(runtimeDefs)
      .map(([key, { type, required, default: defaultDecl }]) => {
        const escapedKey = JSON.stringify(key)
        let defaultString = ''
        if (defaultDecl) {
          defaultString = defaultDecl('default')
        }
        if (!isProduction) {
          return `${escapedKey}: { type: ${toRuntimeTypeString(
            type
          )}, required: ${required}, ${defaultString} }`
        } else if (type.some((el) => el === 'Boolean' || el === 'Function')) {
          return `${escapedKey}: { type: ${toRuntimeTypeString(
            type
          )}, ${defaultString} }`
        } else {
          // production: checks are useless
          return `${escapedKey}: ${
            defaultString ? `{ ${defaultString} }` : 'null'
          }`
        }
      })
      .join(',\n  ')}\n}`

    let decl = runtimeDecls
    if (props.withDefaultsAst && !props.defaults) {
      // dynamic defaults
      decl = `_BD_mergeDefaults(${decl}, ${s.sliceNode(
        props.withDefaultsAst.arguments[1],
        { offset }
      )})`
      // add helper
      s.prependLeft(
        offset,
        `import { mergeDefaults as _BD_mergeDefaults } from 'vue'`
      )
    }
    decl = `defineProps(${decl})`

    s.overwriteNode(props.withDefaultsAst || props.definePropsAst, decl, {
      offset,
    })
  }

  function processEmits(emits: TSEmits) {
    const runtimeDecls = `[${Object.keys(emits.definitions)
      .map((name) => JSON.stringify(name))
      .join(', ')}]`
    s.overwriteNode(emits.defineEmitsAst, `defineEmits(${runtimeDecls})`, {
      offset,
    })
  }
}

function toRuntimeTypeString(types: string[]) {
  return types.length > 1 ? `[${types.join(', ')}]` : types[0]
}
