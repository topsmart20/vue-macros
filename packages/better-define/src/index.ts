import { createUnplugin } from 'unplugin'
import {
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
} from '@vue-macros/common'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import { transformBetterDefine } from './core'
import type { BaseOptions, MarkRequired } from '@vue-macros/common'
import type { UnpluginContextMeta } from 'unplugin'
import type { PluginContext } from 'rollup'

export interface Options extends BaseOptions {
  isProduction?: boolean
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework']
): OptionsResolved {
  const version = options.version || detectVueVersion()

  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC].concat(
      version === 2 && framework === 'webpack' ? REGEX_VUE_SUB : []
    ),
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = 'unplugin-vue-better-define'

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude(id) {
        return filter(id)
      },

      async transform(code, id) {
        try {
          return await transformBetterDefine(code, id, options.isProduction)
        } catch (err: unknown) {
          this.warn(`${name} ${err}`)
          console.warn(err)
        }
      },

      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction
        },

        handleHotUpdate,
      },
    }
  }
)
