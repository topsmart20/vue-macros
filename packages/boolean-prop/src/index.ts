import { type Plugin } from 'vite'
import { type VuePluginApi, getVuePluginApi } from '@vue-macros/common'
import { generatePluginName } from '#macros' assert { type: 'macro' }
import { type Options, transformBooleanProp } from './core/index'

// legacy export
export * from './api'

const name = generatePluginName()

function rollup(options: Options = {}): Plugin {
  let api: VuePluginApi | null | undefined

  return {
    name,
    configResolved(config) {
      try {
        api = getVuePluginApi(config.plugins)
      } catch {}
    },
    buildStart(rollupOpts) {
      if (api === undefined)
        try {
          api = getVuePluginApi(rollupOpts.plugins)
        } catch (error: any) {
          this.warn(error)
          return
        }

      if (!api) return

      api.options.template ||= {}
      api.options.template.compilerOptions ||= {}
      api.options.template.compilerOptions.nodeTransforms ||= []

      api.options.template.compilerOptions.nodeTransforms.push(
        transformBooleanProp(options)
      )
    },
  }
}

export default {
  rollup,
  vite: rollup,
}
