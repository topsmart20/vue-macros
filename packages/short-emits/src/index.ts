import { createUnplugin } from 'unplugin'
import {
  type BaseOptions,
  type MarkRequired,
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  createFilter,
  detectVueVersion,
} from '@vue-macros/common'
import { transformShortEmits } from './core'
import { generatePluginName } from '#macros' assert { type: 'macro' }

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<Options, 'include' | 'version'>

function resolveOption(options: Options): OptionsResolved {
  const version = options.version || detectVueVersion()
  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC],
    ...options,
    version,
  }
}

const name = generatePluginName()

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}) => {
    const options = resolveOption(userOptions)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      transformInclude(id) {
        return filter(id)
      },

      transform(code, id) {
        return transformShortEmits(code, id)
      },
    }
  }
)
