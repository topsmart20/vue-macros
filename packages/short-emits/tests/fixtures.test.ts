import { testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformShortEmits } from '../src/core'

describe('fixtures', async () => {
  await testFixtures(
    import.meta.glob('./fixtures/*.{vue,js,ts}', {
      eager: true,
      as: 'raw',
    }),
    (args, id, code) => transformShortEmits(code, id)?.code
  )
})
