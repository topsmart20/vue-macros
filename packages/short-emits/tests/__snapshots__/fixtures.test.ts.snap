// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > ./fixtures/basic.vue 1`] = `
"<script setup lang=\\"ts\\">
const emit = defineEmits<
  {
    // TSPropertySignature
    (evt: 'input', ...args: []): void
    (evt: 'update:modelValue', ...args: [val: string]): void
    (evt: 'update:color', ...args: [val: number]): void
    (evt: 'change', value: boolean): void

    // TSMethodSignature
    (evt: 'update'): void

    (evt: 'foo\\"\\\\'', value: boolean): void

    (evt: 0, value: boolean): void

    (evt: 'update:modelValue', ...args: [val: string]): void
  }
>()

emit('update:color', 1)
emit('update:modelValue', 'foo')
emit('change', false)
emit(0, false)
</script>
"
`;

exports[`fixtures > ./fixtures/without-se.vue 1`] = `
"<script setup lang=\\"ts\\">
import { defineEmits } from '../../macros' assert { type: 'macro' }

const emit = defineEmits<{
  // TSPropertySignature
  (evt: 'input', ...args: []): void
  (evt: 'update:modelValue', ...args: [val: string]): void
  (evt: 'update:color', ...args: [val: number]): void
  (evt: 'change', value: boolean): void

  // TSMethodSignature
  (evt: 'update'): void

  (evt: 'foo\\"\\\\'', value: boolean): void

  (evt: 0, value: boolean): void
}>()

emit('update:color', 1)
emit('update:modelValue', 'foo')
emit('change', false)
emit(0, false)

const emit2 = defineEmits<(e: 'change') => void>()
emit2('change')
</script>
"
`;
